<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>NGS-pipelines by bigdataage</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">NGS-pipelines</h1>
      <h2 class="project-tagline">Perl source codes for next generation (second generation) sequencing data analysis by integrating lots of open-source softwares.</h2>
      <a href="https://github.com/bigdataage/NGS-Pipelines" class="btn">View on GitHub</a>
      <a href="https://github.com/bigdataage/NGS-Pipelines/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/bigdataage/NGS-Pipelines/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="ngs-pipelines" class="anchor" href="#ngs-pipelines" aria-hidden="true"><span class="octicon octicon-link"></span></a>NGS-Pipelines</h1>

<h2>
<a id="perl-source-codes-for-next-generation-second-generation-sequencing-data-analysis-by-integrating-lots-of-open-source-softwares" class="anchor" href="#perl-source-codes-for-next-generation-second-generation-sequencing-data-analysis-by-integrating-lots-of-open-source-softwares" aria-hidden="true"><span class="octicon octicon-link"></span></a>Perl source codes for next generation (second generation) sequencing data analysis by integrating lots of open-source softwares.</h2>

<p>CISDA: ChIP-Seq Data Analyzer (also contains MNase-seq, DNase-seq, ChIP-exo and all ChIPseq-like data.)                                              </p>

<p>MESDA: MethylC-Seq Data Analyzer    </p>

<p>RASDA: RNA-Seq Data Analyzer                   </p>

<p>All the scripts have been fully tested and passed my quality inspection on Linux Mint, Ubuntu and CentOS.                  </p>

<hr>

<h2>
<a id="the-recent-versions-of-these-softwares-must-be-available-in-your-linux-os" class="anchor" href="#the-recent-versions-of-these-softwares-must-be-available-in-your-linux-os" aria-hidden="true"><span class="octicon octicon-link"></span></a>The recent versions of these softwares must be available in your Linux OS:</h2>

<ol>
<li>Perl, <a href="https://www.perl.org/">https://www.perl.org/</a><br>
</li>
<li>NCBI SRA Toolkit, <a href="http://www.ncbi.nlm.nih.gov/Traces/sra/">http://www.ncbi.nlm.nih.gov/Traces/sra/</a><br>
</li>
<li>FastQC, <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a>
</li>
<li>Trimmomatic, <a href="http://www.usadellab.org/cms/?page=trimmomatic">http://www.usadellab.org/cms/?page=trimmomatic</a><br>
</li>
<li>NGS QC Toolkit, <a href="http://www.nipgr.res.in/ngsqctoolkit.html">http://www.nipgr.res.in/ngsqctoolkit.html</a><br>
</li>
<li>FASTX-Toolkit, <a href="http://hannonlab.cshl.edu/fastx_toolkit/">http://hannonlab.cshl.edu/fastx_toolkit/</a><br>
</li>
<li>Subread, <a href="http://subread.sourceforge.net/">http://subread.sourceforge.net/</a><br>
</li>
<li>BWA, <a href="http://bio-bwa.sourceforge.net/">http://bio-bwa.sourceforge.net/</a><br>
</li>
<li>Bowtie, <a href="http://bowtie-bio.sourceforge.net/index.shtml">http://bowtie-bio.sourceforge.net/index.shtml</a><br>
</li>
<li>Bowtie2, <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">http://bowtie-bio.sourceforge.net/bowtie2/index.shtml</a><br>
</li>
<li>SAMtools, <a href="http://www.htslib.org/">http://www.htslib.org/</a><br>
</li>
<li>BEDtools, <a href="https://github.com/arq5x/bedtools2/releases">https://github.com/arq5x/bedtools2/releases</a><br>
</li>
<li>BEDOPS, <a href="http://bedops.readthedocs.org/en/latest/">http://bedops.readthedocs.org/en/latest/</a><br>
</li>
<li>Deeptools, <a href="http://deeptools.github.io/">http://deeptools.github.io/</a><br>
</li>
<li>MACS1.4, <a href="http://liulab.dfci.harvard.edu/MACS/">http://liulab.dfci.harvard.edu/MACS/</a><br>
</li>
<li>MACS2, <a href="https://github.com/taoliu/MACS/">https://github.com/taoliu/MACS/</a><br>
</li>
<li>HOMER, <a href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a><br>
</li>
<li>DANPOS, <a href="http://code.google.com/p/danpos/">http://code.google.com/p/danpos/</a><br>
</li>
</ol>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/bigdataage/NGS-Pipelines">NGS-pipelines</a> is maintained by <a href="https://github.com/bigdataage">bigdataage</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
