# NGS-Pipelines                          
##Perl source codes for next generation (second generation) sequencing data analysis by integrating lots of open-source softwares.
                                                             
CISDA: ChIP-Seq Data Analyzer                                          
MESDA: MethylC-Seq Data Analyzer                                       
RASDA: RNA-Seq Data Analyzer                   
                                                               
                                                               
  All the scripts have been fully tested and passed my quality inspection on Ubuntu and CentOS.                  
                                               
                                                             
                                                                         
## These softwares (the recent version) must be available in your Linux OS: 
0. Perl, https://www.perl.org/                 
1. NCBI SRA Toolkit, http://www.ncbi.nlm.nih.gov/Traces/sra/                                
2. FastQC, http://www.bioinformatics.babraham.ac.uk/projects/fastqc/
3. Trimmomatic, http://www.usadellab.org/cms/?page=trimmomatic                   
4. NGS QC Toolkit, http://www.nipgr.res.in/ngsqctoolkit.html              
5. FASTX-Toolkit, http://hannonlab.cshl.edu/fastx_toolkit/                  
6. Subread, http://subread.sourceforge.net/               
7. BWA, http://bio-bwa.sourceforge.net/          
8. Bowtie, http://bowtie-bio.sourceforge.net/index.shtml       
9. Bowtie2, http://bowtie-bio.sourceforge.net/bowtie2/index.shtml              
10. SAMtools, http://www.htslib.org/            
11. BEDtools, https://github.com/arq5x/bedtools2/releases                                
12. BEDOPS, http://bedops.readthedocs.org/en/latest/                
13. Deeptools: http://deeptools.github.io/                 




